src/core/qfieldcloud/qfieldcloudproject.cpp

```diff
@@ -1240,90 +1260,95 @@ void QFieldCloudProject::downloadFileConnections( const QString &fileKey )
+  Q_ASSERT( mActiveFilesToDownload.size() == 0 );
+
+  QDir projectPath( QStringLiteral( "%1/%2/%3" ).arg( QFieldCloudUtils::localCloudDirectory(), mUsername, mId ) );
+  const bool currentProjectReloadNeeded = QgsProject::instance()->homePath().startsWith( projectPath.absolutePath() );
+  QStringList gpkgFileNames;
+  if ( currentProjectReloadNeeded )
+  {
+    // we need to close the project to safely flush the gpkg files and avoid file lock on Windows
+    QDirIterator it( projectPath.absolutePath(), { QStringLiteral( "*.gpkg" ) }, QDir::Filter::Files, QDirIterator::Subdirectories );
+    while ( it.hasNext() )
+    {
+      gpkgFileNames << it.nextFileInfo().absoluteFilePath();
+    }
 
-      QDir projectPath( QStringLiteral( "%1/%2/%3" ).arg( QFieldCloudUtils::localCloudDirectory(), mUsername, mId ) );
-      const bool currentProjectReloadNeeded = QgsProject::instance()->homePath().startsWith( projectPath.absolutePath() );
-      QStringList gpkgFileNames;
-      if ( currentProjectReloadNeeded )
+    QgsProject::instance()->clear();
+    if ( mGpkgFlusher )
+    {
+      for ( const QString &fileName : gpkgFileNames )
       {
-        // we need to close the project to safely flush the gpkg files and avoid file lock on Windows
-        QDirIterator it( projectPath.absolutePath(), { QStringLiteral( "*.gpkg" ) }, QDir::Filter::Files, QDirIterator::Subdirectories );
-        while ( it.hasNext() )
-        {
-          gpkgFileNames << it.nextFileInfo().absoluteFilePath();
-        }
+        mGpkgFlusher->stop( fileName );
+      }
+    }
+  }
+
+  // move the files from their temporary location to their permanent one
+  if ( !moveDownloadedFilesToPermanentStorage() )
+  {
+    emit downloadFinished( tr( "Failed to copy some of the downloaded files on your device. Check your device storage." ) );
+    return;
+  }
 
-        QgsProject::instance()->clear();
-        if ( mGpkgFlusher )
+  if ( currentProjectReloadNeeded )
+  {
+    // Clear up Geopackage's shm and wal files
+    for ( const QString &fileName : gpkgFileNames )
+    {
+      QFile shmFile( QStringLiteral( "%1-shm" ).arg( fileName ) );
+      if ( shmFile.exists() )
+      {
+        if ( !shmFile.remove() )
         {
-          for ( const QString &fileName : gpkgFileNames )
-          {
-            mGpkgFlusher->stop( fileName );
-          }
+          QgsMessageLog::logMessage( QStringLiteral( "Failed to remove -shm file '%1' " ).arg( shmFile.fileName() ) );
         }
       }
 
-      // move the files from their temporary location to their permanent one
-      if ( !moveDownloadedFilesToPermanentStorage() )
+      QFile walFile( QStringLiteral( "%1-wal" ).arg( fileName ) );
+      if ( walFile.exists() )
       {
-        emit downloadFinished( tr( "Failed to copy some of the downloaded files on your device. Check your device storage." ) );
-        return;
+        if ( !walFile.remove() )
```

Comment:
I think it would be safer if we check whether the files is empty, no? @nirvn opinion?

Reply:
@suricactus , we want them out no matter what.